doc/design/proposal-ambient-loaders.md

```diff
@@ -0,0 +1,45 @@
+# Ambient Loaders Design
+
+## Problem
+
+Loaders let side logic be applied when resolving import statements. However, this only applies to the main application: loaders themselves aren’t currently affected by any other loaders. As a result, 3rd-party loaders cannot be injected if accessing them requires going through another loader. For instance, the following doesn’t work:
+
+```
+node --loader pnp --loader ts-node ./my-tool.mts
+```
+
+Indeed, importing `ts-node` would require the `pnp` loader to contribute to the resolution, but it’s not what happens today: both `pnp` and `ts-node` are resolved by the default Node resolver, preventing `ts-node` from resolving.
+
+The reason for that is an attempt to keep loaders as isolated as possible, to allow followup improvements like reusing them across workers or scaling them up. If all loaders were influenced by prior loaders, they’d effectively coalesce into a single one, which would prevent such work. Still, the problem remains.
+
+## Proposal
+
+If loaders cannot generally influence each other, we could have a subset of them do. Indeed, not all loaders affect the resolution so much that they are a requirement to followup loaders. We could have two levels of loaders:
+
+- Ambient loaders would be defined via the `--ambient-loader <module>` flag. They would be loaded sequentially, and would affect the resolution of all followup loaders.
```

Comment:
If I registered two ambient loaders, would the first one affect the resolution of the second?

Reply:
Yes.

Reply:
Oh, wait, what. I did not get that from this doc. I understood ambient loaders to work in tandem like how regular loaders do (one regular loader does not affect the other), just specifically when loading a regular loader.

```suggestion
- Ambient loaders would be defined via the `--ambient-loader <module>` flag. They would be loaded sequentially and would affect the resolution of all subsequent ambient loaders.
```

Making ambient loaders affect each other sounds like a pandora's box.

Reply:
Updated with your second wording

> Making ambient loaders affect each other sounds like a pandora's box.

Why that?

Reply:
> Updated with your second wording

Thanks!

>> Making ambient loaders affect each other sounds like a pandora's box.
> Why that?

Wouldn't that open the door for reentry?

Reply:
To be honest I'm still not sure I see a problem with reentrance, we may have different behaviours in mind 😃

But in this case it shouldn't be reentrant anyway: it's just using the result of the previous loader to load the next one, so all execution occurs in the same order and we don't go twice in the same loader during the same resolution pass.

Reply:
What about when one ambient loader depends on another in a circular fashion?

Reply:
Do you have an example?

The closest I can think of would be if, say, we had `/foo.zip/bar.tgz/qux.zip/index.js`, but isn't this pattern true as well for regular module imports?